### DISCUSSION TITLE
Rename DefinePlugin

### DISCUSSION THREAD

**Discussion by matthias-ccri** [webpack/webpack#13571] [Ideas]
Consider renaming "DefinePlugin" to something more accurate. DefinePlugin does not define anything, at least, not in the typical sense of defining a javascript variable.

See examples of places where people think DefinePlugin can create a global variable:
- https://github.com/webpack/webpack/issues/1236
- https://stackoverflow.com/questions/37656592/define-global-variable-with-webpack
- https://stackoverflow.com/questions/41760013/webpack-create-global-variable

What DefinePlugin really does is replace a MemberExpression or Identifier with arbitrary code. It does not define any variable either in the local scope or global scope. Correct me if I'm wrong...

Example:
```js
new webpack.DefinePlugin({
    'ABC': 123,
    'window.ABC': 123,
}),
```
Input:
```js
console.log('ABC' in window);
console.log(window.ABC);
const sss = 'ABC';
console.log(window[sss]);
console.log(window['ABC']);
console.log(ABC);
eval('console.log(ABC)');
```
Output:
```js
console.log('ABC' in window);
console.log(123);
const sss = 'ABC';
console.log(window[sss]);
console.log(123);
console.log(123);
eval('console.log(ABC)');
```
Console:
```
#1 false
#2 123
#3 undefined
#4 123
#5 123
ReferenceError: ABC is not defined
```

This shows that there is no _real_ `ABC` variable being defined anywhere. Many people don't realize that webpack is simply replacing code.

Is it too late to change this plugin's name? Something like `ReplaceVariablesPlugin`? Maybe keep DefinePlugin an alias? 

At the very least, the docs page should revise its description:

> The DefinePlugin allows you to **create global constants** ...

^ this isn't really true.

Anyway, all the best. Let me know what you think. Thanks :+1:

**Comment by alexander-akait** [comment]
We need improve our docs https://github.com/webpack/webpack.js.org/issues/5087

**Reply by ChocolateLoverRaj** [reply]
But the name `DefinePlugin` is more important than the description of it. The name should also be changed / aliased.
